### PR TITLE
Recording after rebase to master

### PR DESCRIPTION
**Pull request by Reflexe**
This is a refactored and rebased-to-master version of #4531 . 

## Bugs
- [x] TripleOscillator preset previews show errors and don't stop due to nonexistent `samples/empty.wav`
- [x] `w2_invsinehalf.flac` has a furry tail in BitInvader due to resampling(maybe an issue with Graph?)
- [x] Visualization gets weird when resizing tracks while recording

## Enhancements

- [ ] (optional) Add more backends
  - [ ] ALSA
  - [ ] CoreAudio
  - [ ] Sndio
  - [ ] libsoundio
- [x] (optional) Refactor
- [x] Decide when to resample data
- [x] Optimize

### CODE REVIEW

File: cmake/nsis/CMakeLists.txt [hunk 1]
```diff
@@ -74,6 +73,27 @@ IF(LMMS_HAVE_STK)
 	INSTALL(FILES ${RAWWAVES} DESTINATION "${DATA_DIR}/stk/rawwaves")
 ENDIF()
 
+# Bundle jack
```
Comment (Reflexe) [inline cmake/nsis/CMakeLists.txt]: Should we drop jack bundling? I was sure it was a good idea back then, right now it does look bizare.

File: include/Mixer.h [hunk 2]
```diff
@@ -315,6 +355,17 @@ class LMMS_EXPORT Mixer : public QObject
 	void requestChangeInModel();
 	void doneChangeInModel();
 
+	RequestChangesGuard requestChangesGuard()
+	{
+		return RequestChangesGuard{this};
+	}
+
+	template<class Callable, class ReturnT=decltype(std::declval<Callable>()())>
+	ReturnT runWhileNotRendering(Callable &&callable) {
```
Comment (necrashter) [inline include/Mixer.h]:
```suggestion
	ReturnT runWhileNotRendering(Callable &&callable)
	{
```
I'm also a curly-brackets-on-the-same-line type of person, but almost all the source files of LMMS puts them in new line. I would recommend doing so for the sake of consistency.

**Comment by PhysSong** [comment]
I think we should close either #4531 or this one in favor of the other one.

**Comment by Reflexe** [comment]
> @Reflexe Just a reminder: do you want to implement or refactor something more at this point?

I want to separate this PR into multiple "easier-to-chew" (test and review) PRs. I just can't find the time for it.

**Comment by PhysSong** [comment]
FYI, I'm on rebasing these changes against current `master`.

**Comment by PhysSong** [comment]
Minor, but If I load a project file with a non-existent sample file, the error message looks like:
```
SoundFile: Could not load: System error: Bad file descriptor
```
On the lastest `master`, it looks like:
```
Sample not found: Spoken.flac
```

**Comment by PhysSong** [comment]
@Reflexe Sorry for repetitively bumping this, but do you have any further plans on this PR? It's one of the features with many demands.

**Comment by Reflexe** [comment]
> @Reflexe Sorry for repetitively bumping this, but do you have any further plans on this PR? It's one of the features with many demands.

Sorry about the delay. I'm actively working on this locally and will keep your all updated.

**Comment by PhysSong** [comment]
The biggest problem of this PR is that it's very hard to review. I think this PR mainly consists of:
- changes in `SampleBuffer` and follow-ups in other classes
- core changes that enable recording
- UI changes for capturing audio
- backend-specific capturing implementations
- build system changes

@Reflexe As you may know, the changes in `SampleBuffer` dominates this PR, making the core functionality hard to review. Since the core functionality heavily depends on those changes, I think you should open a new PR for `SampleBuffer` changes to allow other devs to review it.

**Comment by ChildishGiant** [comment]
Any update on this? I feel like this is a key feature for any DAW

**Comment by PhysSong** [comment]
> I want to separate this PR into multiple "easier-to-chew" (test and review) PRs. I just can't find the time for it.

@Reflexe Do you want/need some help? If it's the case, I can try.

**Comment by therealhammer** [comment]
Just forked and compiled this branch. Great job! Recording worked pretty good with Jack. Looking forward to see this feature in a release version! :)

**Comment by gl3nnleblanc** [comment]
Sorry if I shouldn't bump this, but any update? My workflow would be heavily improved if this is implemented.

> Just forked and compiled this branch. Great job! Recording worked pretty good with Jack. Looking forward to see this feature in a release version! :)

Are there currently any UI elements in place? I built this branch but still see no way to record.

**Comment by qnebra** [comment]
There is a lot more merge conflicts than "non-technical" user can solve.

**Comment by PhysSong** [comment]
As I mentioned a while ago, I'm preparing to revive this PR by splitting to smaller PRs. As far as I understand, this PR consists of:
- Backend changes to support audio capture
- Some changes in core classes like `Mixer`, `SampleRecordHandle`
- `SampleTrack` changes for recording
- Massive `SampleBuffer` refactoring
- Threading-related changes
- UI changes related to recording
- Build system changes(mostly related to JACK on Windows)
- Some CI changes, probably not needed anymore
- Other minor changes

Now I have to analyze their dependency chain and look into individual commits.

**Comment by PhysSong** [comment]
It seems like only a few changes are actually required if we give up supporting live updates while recording. Live updating requires a lot of changes regarding internal data structure, thread safety, and much more.
While trying to understand changes in this PR, I found some odd code fragments. I built and tried to test this PR, but everything was too unstable and buggy to merge this as-is.
I even had to drop a lot of recent commits to get a version which doesn't break much. This is a result of poorly reviewed and tested code.

**Comment by PhysSong** [comment]
Notes for who are looking for downloads: Due to the merge conflicts, neither I or the bot can't provide downloads as-is. As I mentioned above, this pull request is being reworked by me. It's not ready for review at this moment, and there won't be a new pull request until the project reorganization(#5592) finishes.

If there are enough demands for testing binaries, However, I can consider providing unofficial binaries for the new rebased version(Warning: still WIP!).

**Comment by Reflexe** [comment]
I remember trying a few odd things on the latest commits. Removing them should give a relatively working recording functionality. 

Do you have specific instabilities / errors?

**Comment by brigazvi** [comment]
Is there any reason to not merge it to master?

**Comment by IanCaio** [comment]
> Is there any reason to not merge it to master?

That's one of them:

```
This branch has conflicts that must be resolved

Conflicting files
.travis/linux..install.sh
CMakeLists.txt
include/AudioSampleRecorder.h
include/AutomationTrack.h
include/BBTrack.h
include/Engine.h
include/InstrumentTrack.h
include/SampleBuffer.h
include/SampleRecordHandle.h
include/SampleTrack.h
include/SongEditor.h
include/Track.h
include/lmms_basics.h
plugins/audio_file_processor/audio_file_processor.cpp
src/CMakeLists.txt
src/core/Effect.cpp
src/core/Mixer.cpp
src/core/SampleBuffer.cpp
src/core/SamplePlayHandle.cpp
src/core/SampleRecordHandle.cpp
src/core/Song.cpp
src/core/Track.cpp
src/core/audio/AudioDevice.cpp
src/core/audio/AudioJack.cpp
src/core/audio/AudioSdl.cpp
... 
```

**Comment by netpipe** [comment]
it will need to be updated to merge it now ?

**Comment by PhysSong** [comment]
This one is being continued in #5990 and probably more future PRs.

**Comment by Reflexe** [comment]
Moved here: https://github.com/LMMS/lmms/pull/5990

**Pull request closed without merging**